README
What is Facet?

Facet is an unobservable transport for social video sites. Facet evades detection by Internet censors by streaming social videos over Skype calls, and applying a novel traffic-analysis countermeasure called video morphing.

How to use as a client?
1. Add Skype ID of the Facet server into your contact.
2. Send video URLs to the Facet server by Skype messages (or by email).
3. Accept the video call request. 


How to install and run a Facet server?

1. Download Facet 1.0.0
git clone https://github.com/magicle/Facet.git

2. Prerequisite:
(a) Linux operating system, Facet implementation is tested on Ubuntu 12.04
(b) Skype installed with a valid Skype user account
(c) Acknowledgement: Facet server is based on the following open source projects:

Skype4py: a multiplatform Skype API wrapper for Python
v4l2loopback: a kernel module to create V4L2 loopback devices
Gstreamer: a pipeline-based multimedia framework
youtube-dl: command-line program to download videos
Facet repository includes a installer for these projects. 

The "dependency" folder in Facet repository includes a facility to install these required programs.

3. Install by run the following commands:

./configure
make
make install

4. run the server side script:

cd ./src/
python PullMessage.py &		## Skype API
python EmailService.py &	## Extract info by Email




ChangeLog
Change Log;

1. add functionality of receiving address by instant message
2. delete social networking extention facility for simplicity.


dependency/test.sh
dpkg-query -l gstreamer0.10-ffmpeg
echo $?


rasterize.js
var page = require('webpage').create(),
    system = require('system'),
    address, output, size;

var fs = require('fs');
//page.settings.userAgent = 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/28.0.1500.71 Safari/537.36';

page.settings.userAgent = 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/28.0.1500.71 Safari/537.36';

if (system.args.length < 3 || system.args.length > 5) {
    console.log('Usage: rasterize.js URL filename [paperwidth*paperheight|paperformat] [zoom]');
    console.log(' paper (pdf output) examples: "5in*7.5in", "10cm*20cm", "A4", "Letter"');
    console.log(' image (png/jpg output) examples: "1920px" entire page, window width 1920px');
    console.log(' "800px*600px" window, clipped to 800x600');
    phantom.exit(1);
} else {
    address = system.args[1];
    output = system.args[2];
    page.viewportSize = { width: 600, height: 600 };
    if (system.args.length > 3 && system.args[2].substr(-4) === ".pdf") {
        size = system.args[3].split('*');
        page.paperSize = size.length === 2 ? { width: size[0], height: size[1], margin: '0px' }
                                           : { format: system.args[3], orientation: 'portrait', margin: '1cm' };
    } else if (system.args.length > 3 && system.args[3].substr(-2) === "px") {
        size = system.args[3].split('*');
        if (size.length === 2) {
            pageWidth = parseInt(size[0], 10);
            pageHeight = parseInt(size[1], 10);
            page.viewportSize = { width: pageWidth, height: pageHeight };
            page.clipRect = { top: 0, left: 0, width: pageWidth, height: pageHeight };
        } else {
            console.log("size:", system.args[3]);
            pageWidth = parseInt(system.args[3], 10);
            pageHeight = parseInt(pageWidth * 3/4, 10); // it's as good an assumption as any
            console.log ("pageHeight:",pageHeight);
            page.viewportSize = { width: pageWidth, height: pageHeight };
        }
    }
    if (system.args.length > 4) {
        page.zoomFactor = system.args[4];
    }
    page.open(address, function (status) {
        if (status !== 'success') {
            console.log('Unable to load the address!');
            phantom.exit();
        } else {
            window.setTimeout(function () {
                page.render(output);
		fs.write('1.html', page.content, 'w');
                phantom.exit();
            }, 200);
        }
    });
}


src/PullMessage.py
import smtplib
import os
import Skype4Py
import time
import subprocess
import signal
#Skype = Skype4Py.Skype()


global ClientName
global p                            # subprocess
global p_IsInit                     # 1 if p is initiated
global addr
global passcode
global callobj
global cleanbit

# complete  =       0       job exists, initiate call
#           =       1       call in progress
#           =       -1      incomplete, session disrupted
#           =       -2      no job at all
global complete

Skype = Skype4Py.Skype(Transport = 'x11')
Skype.Attach()


if not Skype.Client.IsRunning:
    Skype.Client.Start()


def MailBack(Who, Msg, Subject):
    print "enter"
    fromaddr = 'magiclamp1000@gmail.com'
    toaddrs  = Who
    username = 'magiclamp1000'
    password = '202154215471'

    headers = ["from: "+ fromaddr, "subject: "+ Subject]
    headers = "\r\n".join(headers)

    server = smtplib.SMTP('smtp.gmail.com:587')
    server.ehlo()
    server.starttls()
    server.login(username,password)
    server.sendmail(fromaddr, toaddrs, headers + "\r\n\r\n" + Msg)
    server.quit()
    print "done"



def RequestHandle(User):
    print User.Handle
    f = open("AddFriend", "r+")
    lines = f.readlines()
    for line in lines:
        if User.Handle in line:
            User.BuddyStatus = Skype4Py.enums.budPendingAuthorization
    f.seek(0)
    f.write("")
    f.truncate()
    f.close

def MessageHandle(Message, status):
    skyname = Message.FromHandle

    vurl = Message.Body
    tempf = open ("JobList", 'a')
    tempf.write(skyname + ' ' + vurl + "\n")
    tempf.close()


def clearjob(Skypename):
    flag = 0
    f = open("JobList", "r+")
    lines = f.readlines()
    f.seek(0)
    for i in range(0,len(lines)):
        if Skypename in lines[i] and flag == 0:
                flag = 1
        else:
            f.write(lines[i])
    f.truncate()
    f.close()




def CallHandle(Call, status):
    global p, addr, ClientName, callobj, complete, p_IsInit
    state = Skype.Convert.CallStatusToText(status)
    print state
    ClientName = Call.PartnerHandle
    if "Calling" in state:
        print "Now Start Calling..."
    if "Cancelled" in state:
        complete = -1
    if "Refused" in state:
        complete = -1
    if "Call in Progress" in state:
#        ClientName = Call.PartnerHandle
        print addr
        p = subprocess.Popen(["./play.sh", addr, "/dev/video4"], preexec_fn = os.setsid)
        p_IsInit = 1
        print p.poll()
        print "Connected"
        complete = 1

    if "Finished" in state:
        complete = -1
        print "Waiting for new request"





def GetVurl(url):
    proc = subprocess.Popen(["youtube-dl", "-g", "-f", "36", url], stdout=subprocess.PIPE)
    proc.wait()
    vaddr = proc.stdout.readlines()
    if not vaddr:
        return "ERROR"
    else:
        return vaddr[0].strip("\n")

def DoJob():
    f = open("JobList","rw")
    lines = f.readlines()
    job = "\n"
    for line in lines:
        if len(line.split()[0]) != 0 and job =="\n":
            job = line
    global addr, callobj, complete,ClientName
    f.close()
    if job != "\n":
        ClientName = job.split()[0]
        addr = job.split()[1]
        addr = GetVurl(addr)
        print "addr is:"
        print addr
        if "ERROR" in addr:
            complete = -1
        else:
            callobj = Skype.PlaceCall(job.split()[0])
            complete = 0



Skype.OnCallStatus = CallHandle
Skype.OnMessageStatus = MessageHandle
Skype.OnUserAuthorizationRequestReceived = RequestHandle


flag = 1
while 1:
    if len(Skype.ActiveCalls) == 0:

        #initiate

        p_IsInit = 0
        complete = -2
        flag = 1
        DoJob()
#        print complete
        # do if there exists job...
        if complete != -2:
            while flag:
                time.sleep(1)
                if complete == -1:          #   discrupted
                    flag = 0
                if complete == 1:           #   play ends
                    if p.poll() != None:
                        flag = 0
                # recall 3 times, reject



#            print complete
            # clean to restart

            # end call if not being ended
            if complete == 1:
                callobj.Finish()
            # kill gst process if it's still there
            if p_IsInit == 1:
                if p.poll() == None:
                    os.killpg(p.pid, signal.SIGTERM)
            # delete the job from JobList
            clearjob(ClientName)
    time.sleep(5)














#print len(Skype.ActiveCalls)
#for ch in Skype.ActiveCalls:
#    print ch.PartnerHandle

#Skype.ActiveCalls[0].Answer()





#pp = Skype.SendMessage("skymomo10", "dddd")

#print pp

#for user in Skype.Friends:
#    print '    ', user.FullName




#Skype.PlaceCall("skymomo10")


EmailService.py
# provide email related functionality.

# user send email, whose subject should be: "request", body include:
# "yourskypename".

import time
import email, imaplib, os, random
import smtplib
import subprocess



from email.MIMEMultipart import MIMEMultipart
from email.MIMEBase import MIMEBase
from email import Encoders




global Eaddr,passwd
Eaddr = "magiclamp1000@gmail.com"
passwd = "202154215471"

def Branch(mail):
    if mail["Subject"] == "search":
        VideoSearch(mail)
        return 1
    if mail["Subject"] == "url":
        UrlHandle(mail)
        return 1
    elif mail["Subject"] == "myemail":
        MyemailHandle(mail)
    elif mail["Subject"] == "add":
        AddHandle(mail)
        return 0

def VideoSearch(mail):
    query = BodyOneLine(mail)
    Url = "https://www.google.com/search?tbm=vid&q="
    for each_term in query.split(" "):
        Url = Url + "+" + each_term
    print Url
    proc = subprocess.Popen(["phantomjs", "rasterize.js", Url, "temp.pdf"])
    proc.wait()
    print "search finished"
    MailAttach(mail["From"], "Search Result", "temp.pdf")
    print "PDF file has been sent!"

def AddHandle(mail):
    f = open("AddFriend", "r+")
    old = f.read()
    f.seek(0)
    left = mail["From"].index("<")+1
    right = mail["From"].index(">")
    package = mail["From"][left:right] + " " + BodyOneLine(mail) + "\n"
    f.write(package + old)
    f.truncate()
    f.close()
    MailBack(mail["From"][left:right], BodyOneLine(mail) + "has been allowed to join the service! Please let him add server into his contact list!", "Adding friend succeeds!")

def MyemailHandle(mail):
    f = open("Pending","r+")
    old = f.read()
    f.seek(0)
    left = mail["From"].index("<") + 1
    right = mail["From"].index(">")
    source = mail["From"][left:right]
    Skyname = BodyOneLine(mail).split()[0]
    Passcode = BodyOneLine(mail).split(" ", 1)[1]
    f.write(Skyname + " " + source + " " + Passcode + "\n" + old)
    f.close
    MailBack(mail["From"], "Success!", "Pending!")



def EmailCheck(mail):
    left = mail["From"].index("<")
    right = mail["From"].index(">")
    source = mail["From"][left + 1:right]

    f= open("EmailCheck","r")
    lines = f.readlines()
    for line in lines:
        if source in line:
            return line.split(" ")[0]
    return 0


def UrlHandle(mail):
    name = EmailCheck(mail)
    if name != 0:
        f= open("JobList", "wa")
        print BodyOneLine(mail)
        f.write(name + " " + BodyOneLine(mail) + "\n")
        f.close
        MailBack(mail["From"], "Success!","On Queue!")
    else:
        MailBack(mail["From"], "Wrong!", "Please Register!")

def BodyOneLine (msg):
    maintype = msg.get_content_maintype()
    if maintype == 'multipart':
        for index, part in enumerate(msg.walk()):
          if index == 1:
            return part.get_payload().split("\n")[0]

    elif maintype == 'text':
        return msg.get_payload().split("\n",1)[0]



def MailAttach(Who, Subject, Atpath):

    global Eaddr, passwd
    fromaddr = Eaddr
    toaddrs = Who
    username = Eaddr.split("@")[0]
    password = passwd

    msg = MIMEMultipart()
    msg['Subject'] = Subject
    msg['From'] = fromaddr
#    msg['To'] = ', '.join(toaddrs)
    msg['To'] = toaddrs

    part = MIMEBase('application', "octet-stream")
    part.set_payload(open(Atpath, "rb").read())
    Encoders.encode_base64(part)

    part.add_header('Content-Disposition', 'attachment; filename=' + Atpath)
    msg.attach(part)

    server = smtplib.SMTP('smtp.gmail.com:587')
    server.ehlo()
    server.starttls()
    server.login(username,password)
    server.sendmail(fromaddr, toaddrs, msg.as_string())
    server.quit()



def MailBack(Who, Msg, Subject):
    global Eaddr, passwd
    fromaddr = Eaddr
    toaddrs  = Who
    username = Eaddr.split("@")[0]
    password = passwd

    headers = ["from: "+ fromaddr, "subject: "+ Subject]
    headers = "\r\n".join(headers)

    server = smtplib.SMTP('smtp.gmail.com:587')
    server.ehlo()
    server.starttls()
    server.login(username,password)
    server.sendmail(fromaddr, toaddrs, headers + "\r\n\r\n" + Msg)
    server.quit()



def init():
    if not os.path.exists("Pending"):
        f = open("Pending", "w")
        f.write("\n")
        f.close
    if not os.path.exists("EmailCheck"):
        f= open("EmailCheck", "w")
        f.write("\n")
        f.close
    if not os.path.exists("JobList"):
        f = open("JobList", "w")
        f.write("\n")
        f.close
    if not os.path.exists("AddFriend"):
        f = open("AddFriend", "w")
        f.write("\n")
        f.close
    return 1

init()
while 1:
    time.sleep(2)
    detach_dir = '.' # directory where to save attachments (default: current)
    user = Eaddr.split("@")[0]
    pwd = passwd

# connecting to the gmail imap server
    m = imaplib.IMAP4_SSL("imap.gmail.com")
    m.login(user,pwd)
    m.select("[Gmail]/All Mail") # here you a can choose a mail box like INBOX instead
# use m.list() to get all the mailboxes


    resp, items = m.search(None, "UNSEEN") # you could filter using the IMAP rules here (check http://www.example-code.com/csharp/imap-search-critera.asp)

    items = items[0].split() # getting the mails id

    for emailid in items:
        resp, data = m.fetch(emailid, "RFC822") # fetching the mail, "`(RFC822)`" means "get the whole stuff", but you can ask for headers only, etc
        email_body = data[0][1] # getting the mail content
        mail = email.message_from_string(email_body) # parsing the mail content to get a mail object

        print mail["Subject"]


        Branch(mail)

#    print showme(mail)
